Add authentication for rate and geo services for SREBench fault injections.

HacksonClark opened the pull request.

Commit:
Add authentication for rate and geo services for SREBench authentication faults.

menidd commented on hotelReservation/helm-chart/hotelreservation/templates/_baseDeploymentMongoDB.tpl:
```diff
@@ -22,74 +22,27 @@ spec:
-        resources:
-          {{ tpl $.Values.global.resources $ | nindent 10 | trim }}
+        - "--auth"
         {{- end }}
```
I understand the new stuff, and changing the `{{- with .Values.container }}` is a matter of taste, but the `imagePullPolicy` and `ports` block look important, as do a number of the sections removed below including `affinity`.  Why isn't this just a new `--auth` option and ConfigMap mount?

menidd commented on hotelReservation/helm-chart/hotelreservation/templates/configs/service-config.tpl:
```diff
@@ -4,12 +4,12 @@
     "FrontendPort": "5000",
     "GeoPort": "8083",
-    "GeoMongoAddress": "mongodb-geo-{{ include "hotel-reservation.fullname" . }}.{{ .Release.Namespace }}.svc.{{ .Values.global.serviceDnsDomain }}:27018",
+    "GeoMongoAddress": "admin:admin@mongodb-geo-{{ include "hotel-reservation.fullname" . }}.{{ .Release.Namespace }}.svc.{{ .Values.global.serviceDnsDomain }}:27018",
```
Shouldn't these both be present with an if/else on `.Values.useAccessControl`?  (Same goes for line 12 below.)

menidd commented on hotelReservation/helm-chart/hotelreservation/charts/mongodb-rate/values.yaml:
```diff
@@ -12,3 +12,9 @@ container:
   ports:
   - containerPort: 27017
+  args:
+    - "--auth" # MongoDB access control failure case
```
This argument is added in `_baseDeploymentMongoDB.tpl` when `.Values.useAccessControl` is set.  It should not be hard-coded in the `args` list.

menidd commented on hotelReservation/helm-chart/hotelreservation/charts/mongodb-rate/templates/mongo-rate-configmap.yaml:
```diff
@@ -0,0 +1,45 @@
+apiVersion: v1
```
This whole file should be wrapped in an `if .Values.useInitScript`

HacksonClark replied:
I'll do this, but I'm not sure how much it matters if we only apply the configmap if Values.useInitScript is true.

menidd replied:
It adds a noise artifact that may distract analysis.  This might be a valid challenge (simulating a badly managed cluster), but only if we do it intentionally as part of a specific issue.  The normal incident environments should be as clean as possible.

menidd commented on hotelReservation/helm-chart/hotelreservation/charts/mongodb-geo/values.yaml:
```diff
@@ -12,3 +12,9 @@ container:
+  args:
+    - "--auth" # MongoDB access control failure case
+
+useAccessControl: true
+useInitScript: true
+initScriptConfigMap: mongo-geo-script
```
These are set again in `mongodb-rate` and will need to match.  Shouldn't they be instead in the overall `values`?

HacksonClark replied:
@menidd i don't think so, we're only using access control on two services (geo and rate) so there isn't a need for authentication overall.

menidd replied:
It's currently set in `geo`, `mongodb-geo`, and `mongodb-rate`.  Under what conditions would these three places not all need to set it to the same value?  If the answer is "never," then wouldn't it be better to have a global `enableMongoAuthentication`?  Also, I still don't see the value of a separate `useInitScript` and `useAccessControl`; when will those correctly be set to different values?

menidd reviewed (changes requested):
This will be a good place to introduce issues.  Incorrect passwords are a common cause of failure.

As it is, this will change the normal behaviour also when authentication is not enabled.  
Since it needs to be known in multiple containers that authentication is required, the flag should be global rather than individually set in the individual deployment values, and it should be set to `false` so the helm deployment can optionally override it to `true`.  It's probably not worth two flags for enabling authentication and using a mongo init script; if an init script is required for another case, then it will be a different script anyhow.

It's not clear why a large part of `_baseDeploymentMongoDB.tpl` has been removed.  Some of the deleted blocks look useful.

Commit:
Remove auth hardcoding as it's unneccessary.

Commit:
Add in helm options that were previously dropped.

Commit:
Added conditional to configmaps.

Commit:
Update service config to use conditionals.

HacksonClark committed:
Merge branch 'master' into mongo-auth

HacksonClark commented:
@menidd I believe I've addressed all of your comments with my latest commits. Let me know if you have anymore suggestions and thank you for all your help so far.

menidd commented on hotelReservation/helm-chart/hotelreservation/templates/_baseDeploymentMongoDB.tpl:
```diff
@@ -31,65 +31,71 @@ spec:
+          name: {{ .Values.initScriptConfigMap }}
       {{- end }}
-      hostname: {{ .Values.name }}-{{ include "hotel-reservation.fullname" . }}
-      restartPolicy: {{ .Values.restartPolicy | default .Values.global.restartPolicy}}
```
Is there a reason for dropping these?  `hostname` is very helpful, and `restartPolicy` could easily be part of an incident scenario, but they are being dropped from all charts (that is, not just with an if).  Do they conflict with something, or are they made redundant by some other addition?

menidd commented on hotelReservation/helm-chart/hotelreservation/templates/_baseDeploymentMongoDB.tpl:
```diff
@@ -31,65 +31,71 @@ spec:
+      nodeSelector: {{ toYaml .Values.nodeSelector | nindent 8 }}
       {{- else if hasKey $.Values.global "nodeSelector" }}
-      nodeSelector: {{- toYaml .Values.global.nodeSelector | nindent 8 }}
+      nodeSelector: {{ toYaml $.Values.global.nodeSelector | nindent 8 }}
```
Is there a reason for replacing the `{{-` with `{{` for all these `toYaml` blocks?  Those are very picky with indentation, so if they worked before then it's dangerous to change them without testing of each one.

menidd commented on hotelReservation/helm-chart/hotelreservation/charts/mongodb-rate/values.yaml:
```diff
@@ -12,3 +12,7 @@ container:
   ports:
   - containerPort: 27017
+
+useAccessControl: true
+useInitScript: true
```
These should default false.  Like this, all legacy configurations will suddenly start deploying with the new feature enabled.

menidd commented on hotelReservation/helm-chart/hotelreservation/charts/mongodb-rate/values.yaml:
```diff
@@ -12,3 +12,7 @@ container:
+
+useAccessControl: true
+useInitScript: true
+initScriptConfigMap: mongo-rate-script
```
Why is the script ConfigMap name a configuration variable?

HacksonClark replied:
We use a different script for each service, so in order for the base deployment to use the right one it's a variable.

menidd replied:
I see that now in geo and rate.  Thanks.

menidd reviewed (changes requested):
Thanks for the updates.  I still see some potential problems though.  The main background for my thoughts:
1. There should be as few configuration variables as possible, so if two are always changed in synchronization, they should be a single variable.
2. If the new configuration variables are not used, the charts should deploy exactly as before.

I'm happy to listen to arguments for any of these changes, and I have tried to present my reasons in the comments, so if you disagree then please indicate your reasons in the conversations.

If a more interactive discussion would help, then I'm also happy to have a video meeting with you, me, and @saurabhjha1 .  (Early, please, I'm in CET, seven hours later than Illinois.)

Commit:
Added back hostname and restartPolicy.

Commit:
Fixed toYaml blocks.

Commit:
Removed useInitScript variable as it's redundant with useAccessControl.

Commit:
Default useAccessControl to false.

HacksonClark commented:
@menidd I just pushed in some changes that address your comments. I am currently having an issue with the tlsCertificates after pulling:

`TASK [sample_applications : Deploy Deathstar Hotel Reservations] **************************************************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "command": "/usr/local/bin/helm upgrade -i --reset-values --create-namespace --set global.monitoring.otelAddress=otel-collector-hotel-reservations.hotel-reservations.svc.cluster.local --set global.monitoring.defaultJaegerEnabled=false --set global.monitoring.centralJaegerAddress=jaeger-collector.jaeger.svc.cluster.local --set global.monitoring.centralPrometheusAddress=http://prometheus-stack-kube-prom-prometheus.prometheus:9090/api/v1/otlp --set global.nameOverride=hotel-reservations hotel-reservations 'roles/sample_applications/DeathStarBench/hotelReservation/helm-chart/hotelreservation'", "msg": "Failure when executing Helm command. Exited 1.\nstdout: Release \"hotel-reservations\" does not exist. Installing it now.\n\nstderr: Error: template: hotel-reservation/charts/wrk2/templates/deployment.yaml:1:3: executing \"hotel-reservation/charts/wrk2/templates/deployment.yaml\" at <include \"hotelreservation.templates.wrk2-deployment\" .>: error calling include: template: hotel-reservation/templates/wrk2_deployment.tpl:3:7: executing \"hotelreservation.templates.wrk2-deployment\" at <eq $.Values.global.services.environments.TLS 1>: error calling eq: incompatible types for comparison\n", "stderr": "Error: template: hotel-reservation/charts/wrk2/templates/deployment.yaml:1:3: executing \"hotel-reservation/charts/wrk2/templates/deployment.yaml\" at <include \"hotelreservation.templates.wrk2-deployment\" .>: error calling include: template: hotel-reservation/templates/wrk2_deployment.tpl:3:7: executing \"hotelreservation.templates.wrk2-deployment\" at <eq $.Values.global.services.environments.TLS 1>: error calling eq: incompatible types for comparison\n", "stderr_lines": ["Error: template: hotel-reservation/charts/wrk2/templates/deployment.yaml:1:3: executing \"hotel-reservation/charts/wrk2/templates/deployment.yaml\" at <include \"hotelreservation.templates.wrk2-deployment\" .>: error calling include: template: hotel-reservation/templates/wrk2_deployment.tpl:3:7: executing \"hotelreservation.templates.wrk2-deployment\" at <eq $.Values.global.services.environments.TLS 1>: error calling eq: incompatible types for comparison"], "stdout": "Release \"hotel-reservations\" does not exist. Installing it now.\n", "stdout_lines": ["Release \"hotel-reservations\" does not exist. Installing it now."]}`

I was wondering if you had any ideas why that might be happening. All the code I pushed I did test successfully, but the current state of the branch isn't working for me after pulling.

menidd commented on hotelReservation/helm-chart/hotelreservation/templates/_baseDeploymentMongoDB.tpl:
```diff
@@ -31,65 +31,73 @@ spec:
-        image: {{ .dockerRegistry | default $.Values.global.dockerRegistry }}/{{ .image }}:{{ .imageVersion | default $.Values.global.defaultImageVersion }}
-        imagePullPolicy: {{ .imagePullPolicy | default $.Values.global.imagePullPolicy }}
+      - name: "{{ .Values.container.name }}"
+        image: "{{ .Values.container.image }}:{{ .Values.container.imageVersion }}"
+        imagePullPolicy: {{ .Values.container.imagePullPolicy | default $.Values.global.imagePullPolicy }}
```
Is there a reason the `.dockerRegistry` value is no longer being used?  Also, the default value for `.imageVersion` has been dropped.

HacksonClark replied:
The old code was just causing this error for me:
`Deployment.apps "mongodb-user-hotel-reservations" is invalid: spec.template.spec.containers[0].name: Required value
`

We have the information we need in the values.yaml files which is why we use it, though maybe we need to fix it at a higher level.

menidd replied:
While processing `charts/mongodb-user/values.yaml` with the `{{- with .Values.container }}` it did not pick up the `{{ .name }}`, but changing it to `{{ .Values.container.name }}` without the with block worked?  That's worth understanding eventually, but if it fixed it then OK.  Changing the `image` value shouldn't have affected the `name`, although I notice the old version does seem to be missing the double-quotes around the values for both `image` and `imagePullPolicy`, which might have been a problem for some values.

menidd commented on hotelReservation/helm-chart/hotelreservation/templates/_baseDeploymentMongoDB.tpl:
```diff
@@ -31,65 +31,73 @@ spec:
-        {{- range $cport := .ports }}
-        - containerPort: {{ $cport.containerPort -}}
-        {{ end }}
-        {{- if .command}}
-        command:
-        - {{ .command }}
-        {{- end -}}
```
The `command` block has been dropped.

menidd commented on hotelReservation/helm-chart/hotelreservation/templates/_baseDeploymentMongoDB.tpl:
```diff
@@ -31,65 +31,73 @@ spec:
+      {{- end }}
+      {{- if hasKey .Values "topologySpreadConstraints" }}
+      topologySpreadConstraints:
+        {{ tpl .Values.topologySpreadConstraints . | nindent 6 }}
+      {{- else if hasKey $.Values.global.mongodb "topologySpreadConstraints" }}
+      topologySpreadConstraints:
+        {{ tpl $.Values.global.mongodb.topologySpreadConstraints . | nindent 6 }}
```
a `| trim` has also been dropped from both of the `nindent 6` lines.

menidd reviewed (changes requested):
Thanks for your help and patience.  These are my last comments, but I'm also not going to be very responsive for the next couple of weeks.  @saurabhjha1 this looks good to me (with these last fixes); are you around to click the approve later this week?
